<a href="https://colab.research.google.com/github/izzat-ai/learning-ai/blob/main/projects/diabet_project/diabetes_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [35]:
df = pd.read_csv("https://raw.githubusercontent.com/anvarnarz/praktikum_datasets/refs/heads/main/diabetes.csv")
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


Datasetimizdagi ustunlar :

- Pregnancies - homilador bo'lish soni
- Glucose - glyukozaga(shakar miqdori) test natijasi
- BloodPressure - diastolik qon bosimi (mm Hg)
- SkinThickness Triceps - teri burmasining qalinligi (mm)
- Insulin - 2 soatlik sarum insulini (mu U/ml)
- BMI - Tana massasi indeksi (vazn kg / (m bo'yi) ^ 2)
- DiabetesPedigreeFunction - diabetning naslchilik funktsiyasi
- Age - Yosh (yil)
- Outcome - Class (0 - diabet yo'q, 1 - diabet)



In [37]:
df.shape

(768, 9)

In [38]:
# diabet ma'lumotlarida 0 qiymatlar ko'p uchrashi mumkin va ular tabiiy bo'miydi
(df==0).sum()

,0
Pregnancies,111
Glucose,5
BloodPressure,35
SkinThickness,227
Insulin,374
BMI,11
DiabetesPedigreeFunction,0
Age,0
Outcome,500


- Outcome (target) ustunida 0 lar bo'lishi tabiiy , chunki u faqat 1 va 0 dan iborat
- Pregnancies ustunidagi 0 lar ham tabiiy bo'ladi , chunki homilador bo'magan bo'lishi mumkin
- Glucose = 0 , BloodPressure = 0 , SkinThickness = 0 , Insulin = 0 , BMI = 0 bo'lishi hayotdagi be'morlarda bo'lishi imkonsiz . Shuning uchun bu 0 lar NaN o'rniga to'ldirilgan deb qaraladi

In [39]:
df.describe()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
count,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000,768.000000
mean,3.845052,120.894531,69.105469,20.536458,79.799479,31.992578,0.471876,33.240885,0.348958
std,3.369578,31.972618,19.355807,15.952218,115.244002,7.884160,0.331329,11.760232,0.476951
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.078000,21.000000,0.000000
25%,1.000000,99.000000,62.000000,0.000000,0.000000,27.300000,0.243750,24.000000,0.000000
50%,3.000000,117.000000,72.000000,23.000000,30.500000,32.000000,0.372500,29.000000,0.000000
75%,6.000000,140.250000,80.000000,32.000000,127.250000,36.600000,0.626250,41.000000,1.000000
max,17.000000,199.000000,122.000000,99.000000,846.000000,67.100000,2.420000,81.000000,1.000000


In [40]:
# target ustunimizni balansini teksirish
df['Outcome'].value_counts()

,count
Outcome,
0,500
1,268


- 0 - sog'lom odamlar 500 ta , diabetga chalingan odamlar esa 268 ta
- ushbu datasetimiz biroz disbalans

In [41]:
# duplikatlarni aniqlash
df.duplicated().sum()

np.int64(0)

In [42]:
# ustun nomlarini tog'ri yozilganini tekshirish
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [44]:
# 0 bo'lishi mumkin bo'magan ustunlarni 0 larini NaN qilish
zeros_cols = ['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']
df[zeros_cols] = df[zeros_cols].replace(0, np.nan)
df.isnull().sum()

,0
Pregnancies,0
Glucose,5
BloodPressure,35
SkinThickness,227
Insulin,374
BMI,11
DiabetesPedigreeFunction,0
Age,0
Outcome,0


In [46]:
# 0 lar qomaganini tekshirish
(df == 0).sum()

,0
Pregnancies,111
Glucose,0
BloodPressure,0
SkinThickness,0
Insulin,0
BMI,0
DiabetesPedigreeFunction,0
Age,0
Outcome,500


In [47]:
# NaN lar qanchaligini foizda ko'rish
(df.isnull().sum()/len(df)*100).round(2)

,0
Pregnancies,0.00
Glucose,0.65
BloodPressure,4.56
SkinThickness,29.56
Insulin,48.70
BMI,1.43
DiabetesPedigreeFunction,0.00
Age,0.00
Outcome,0.00


- Insulin va SkinThickness ustunlaridagi NaN qiymatlar judayam ko'p . Qolgan ustunlarda unchalik ko'p emas

###**Ma'lumotlarni train va test setlarga bo'lish**

In [48]:
# features va target labelni ajratish
X = df.drop("Outcome", axis=1)
y = df["Outcome"]

from sklearn.model_selection import train_test_split
# train va testlarga bo'lish | targetdagi balansni saqlagan holda
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [50]:
print(X_train.shape)
print(X_test.shape)

(614, 8)
(154, 8)


In [51]:
print(y_train.value_counts())
print(y_test.value_counts())

Outcome
0    400
1    214
Name: count, dtype: int64
Outcome
0    100
1     54
Name: count, dtype: int64


In [52]:
# NaN qiymatlar bor edi , ularni tekshiramiz
X_train.isnull().sum()

,0
Pregnancies,0
Glucose,4
BloodPressure,23
SkinThickness,175
Insulin,290
BMI,9
DiabetesPedigreeFunction,0
Age,0


- Insulin va SkinThickness ustunlarini median qiymat bilan to'ldirsak , juda ko'p bir xil qiymatlar takrorlangan bo'ladi , ma'lumotlarni sun'iy ko'paytirgan bo'lamiz
- Glucose, BloodPressure, BMI ustunlari uchun esa median qiymat bilan to'ldirsak hech qanday kamchilik kuzatilmiydi

# Glucose, BloodPressure, SkinThickness, Insulin va BMI ustunlaridagi 0 larni nan qilish
df['Glucose'] = df['Glucose'].replace(0, np.nan)
df['BloodPressure'] = df['BloodPressure'].replace(0, np.nan)
df['SkinThickness'] = df['SkinThickness'].replace(0, np.nan)
df['Insulin'] = df['Insulin'].replace(0, np.nan)
df['BMI'] = df['BMI'].replace(0, np.nan)

# nan qiymatlar sonini aniqlash
df.isnull().sum()

### Train_test split